import pandas as pd
import os
from snakemake.utils import min_version
from snakemake.shell import shell

# Minimum Snakemake version
min_version("8.8.0")

# Load config
configfile: "config/config.yaml"

# Load sample sheet
samplesheet_df = pd.read_csv(config["sample_sheet"])
samplesheet_df["genome"] = (
    samplesheet_df["genome"].astype(str).str.strip("\"' \t")
)
samples = samplesheet_df["sample_id"].tolist()

# Ensure only one genome is used
genomes = samplesheet_df["genome"].unique().tolist()
if len(genomes) > 1:
    raise ValueError("Multiple genome types detected! Use only one genome in the sample sheet.")
selected_genome = genomes[0]  # e.g. "hg38" or "mm39"

# Load comparison list from CSV
comparisons_file = "data/comparison.csv"
comparisons_df = pd.read_csv(comparisons_file)
comparisons = comparisons_df.apply(
    lambda row: f"{row['treatment']}_vs_{row['control']}",
    axis=1
).tolist()

# Final report output
report: "report/workflow.rst"

# Always include BOTH rule sets; Snakemake will only run the rules needed
include: "rules/download_ref_hg38.smk"
include: "rules/download_ref_mm39.smk"
include: "rules/convert_gtf_to_bed.smk"

# Additional reference-based annotation
include: "rules/create_rseqc_annotation.smk"

##################################
# QC and Trimming
##################################
include: "rules/download_atria.smk"
include: "rules/detect_adapters.smk"
include: "rules/fastqc_raw.smk"
include: "rules/multiqc_raw.smk"
include: "rules/trimming.smk"
include: "rules/fastqc_trimmed.smk"
include: "rules/multiqc_trimmed.smk"

##################################
# Alignment and Quantification
##################################
include: "rules/alignment_star.smk"
include: "rules/index_bam.smk"
include: "rules/samtools_stats.smk"
include: "rules/mosdepth.smk"
include: "rules/multiqc_bam.smk"
include: "rules/quantify_transcripts_salmon.smk"
include: "rules/combine_expression_matrix.smk"
include: "rules/transcript_to_gene_mapping.smk"
include: "rules/merge_samtools_stats.smk"
include: "rules/extract_bam_stats.smk"
include: "rules/generate_align_pct.smk"
include: "rules/gene_body_coverage.smk"
include: "rules/infer_experiment.smk"
include: "rules/read_distribution.smk"
include: "rules/read_duplication.smk"
include: "rules/rpkm_saturation.smk"
include: "rules/junction_saturation.smk"
include: "rules/bam_stat.smk"
include: "rules/process_genome_content.smk"
include: "rules/go_enrichment.smk"
include: "rules/deseq_analysis_comparison.smk"

rule all:
    input:
        "../results/Quant/Count/combined_expression.csv",
        expand("../results/Quant/Count/quant/{sample}/quant.sf", sample=samples),
        expand("../results/Differential/deglist/{comparison}_deg_results.csv", comparison=comparisons),
        "../results/Quant/transcript_to_gene_mapping.csv",
        expand("../rawdata/{sample}_{suffix}.fq.gz", sample=samples, suffix=["1", "2"]),
        expand("../results/adapters/{sample}_adapters.txt", sample=samples),
        expand("../results/trimmed/{sample}_{suffix}.atria.fq.gz", sample=samples, suffix=["1", "2"]),
        expand("../results/fastqc/trimmed/{sample}_{suffix}.atria_fastqc.html", sample=samples, suffix=["1", "2"]),
        expand("../results/fastqc/raw/{sample}_{suffix}_fastqc.html", sample=samples, suffix=["1", "2"]),
        "../results/multiqc/trimmed/multiqc_report.html",
        "../results/multiqc/raw/multiqc_report.html",
        "../results/multiqc/bam/multiqc_report.html",
        expand("../results/aligned/{sample}.bam", sample=samples),
        expand("../results/aligned/{sample}.bam.bai", sample=samples),
        expand("../results/qc/samtools/{sample}_samtools_stats.txt", sample=samples),
        expand("../results/qc/samtools/{sample}_flagstat.txt", sample=samples),
        expand("../results/qc/rseqc/{sample}_infer_experiment.txt", sample=samples),
        expand("../results/qc/rseqc/{sample}_read_distribution.txt", sample=samples),
        expand("../results/qc/rseqc/{sample}.geneBodyCoverage.txt", sample=samples),
        expand("../results/qc/rseqc/{sample}.geneBodyCoverage.curves.pdf", sample=samples),
        expand("../results/qc/rseqc/{sample}.junctionSaturation_plot.pdf", sample=samples),
        expand("../results/qc/rseqc/{sample}.DupRate_plot.pdf", sample=samples),
        expand("../results/qc/rseqc/{sample}.DupRate_plot.r", sample=samples),
        expand("../results/qc/rseqc/{sample}.saturation.pdf", sample=samples),
        expand("../results/qc/rseqc/{sample}.saturation.r", sample=samples),
        expand("../results/qc/rseqc/{sample}.eRPKM.xls", sample=samples),
        expand("../results/qc/rseqc/{sample}_read_distribution.txt", sample=samples),
        expand("../results/qc/rseqc/{sample}_bam_stat.txt", sample=samples),
        expand("../results/qc/mosdepth/{sample}.mosdepth.summary.txt", sample=samples),
        expand("../results/enrichment/GO/Biological_Process/{comparison}_biological_process_enrichment.csv", comparison=comparisons),
        expand("../results/enrichment/GO/Molecular_Function/{comparison}_molecular_function_enrichment.csv", comparison=comparisons),
        expand("../results/enrichment/GO/Cellular_Component/{comparison}_cellular_component_enrichment.csv", comparison=comparisons),

        # The reference files + Salmon index for the selected genome
        f"../results/genome/{selected_genome}/{selected_genome}_annotation.gtf",
        f"../results/genome/{selected_genome}/{selected_genome}_genome.fa",
        f"../results/genome/{selected_genome}/{selected_genome}_transcriptome.fa",
        f"../results/genome/{selected_genome}/salmon/salmon_index_{selected_genome}",
        f"../results/genome/{selected_genome}/{selected_genome}.bed",
        "../results/tools/Atria/atria-4.1.1/bin/atria"
