# bulk_RNA_seq
Snakemake pipeline processing fastq files to the DESeq2, and GO annotations

## Pipeline Features

- **FastQ Quality Control** (`FastQC`, `MultiQC`)
- **Adapter Trimming** (`Atria`)
- **Read Alignment** (`STAR`)
- **Transcript Quantification** (`Salmon`)
- **Quality Control Metrics** (`RSeQC`, `samtools`)
- **Differential Expression Analysis** (`DESeq2`)
- **GO Enrichment Analysis** (`clusterProfiler`)

## üì• Installation

### 1Ô∏è‚É£ Install Dependencies

```bash
conda install -c bioconda snakemake
```

### 2Ô∏è‚É£ Clone the Repository
```bash
git clone https://github.com/your_username/Bulk_RNA_seq.git
cd Bulk_RNA_seq
```

## üöÄ Running the Pipeline

### 1Ô∏è‚É£ Test Pipeline Execution (Dry Run)
```bash
snakemake --use-conda -np
```

### 2Ô∏è‚É£ Run the Full Pipeline
```bash
snakemake --use-conda --cores 30
```
üí° *Adjust `--cores` based on your system's available CPUs.*

### 3Ô∏è‚É£ Run a Specific Step
```bash
snakemake --use-conda --cores 30 salmon_quant_reads
```

### 4Ô∏è‚É£ Force Rerun of a Specific Step
```bash
snakemake --use-conda -R build_salmon_index --cores 30
```

### 5Ô∏è‚É£ Rerun If File Modification Time Has Changed (recommended)

If you have input files in correct format, you should skip generating them again by: 

```bash
snakemake --use-conda --rerun-triggers mtime --cores 30
```

